### ISSUE TITLE
Pathologically slow Hexagon codegen for certain schedules

### ISSUE THREAD

**Issue by steven-johnson** [halide/Halide#4735]
Steps to repeat. We are going to build and run one of the Generators in `apps/nn_ops`:

```
$ cd apps/nn_ops
$ make bin/host/DepthwiseConvolution.generator
$ time bin/host/DepthwiseConvolution.generator -g DepthwiseConvolution -o /tmp -e object -f DepthwiseConvolution_1 target=x86-64-linux-hvx_128-no_runtime depth_multiplier=1
```

At current master branch, on my stupid fast Linux box, I get:

```
real  0m9.602s
user  0m9.528s
sys 0m0.078s
```

To see the pathological schedule, go to `apps/nn_ops/DepthwiseConvoution_generator.cpp` and change the lines

```
        output_.split(y, y, yi, y_split_factor).parallel(y);
        output_.vectorize(depth, vector_size_u8, TailStrategy::RoundUp);
```

to

```
        Var di("di"), xi("xi");
        output_.split(y, y, yi, y_split_factor, TailStrategy::GuardWithIf)
            .tile(depth, x, di, xi, vector_size_u8, 2, TailStrategy::GuardWithIf)
            .vectorize(di)
            .unroll(xi)
            .parallel(y);
```

Now, re-run the steps above (still on `master`). I get

```
real  45m51.237s
user  45m37.659s
sys 0m13.206s
```

**Comment by abadams** [comment]
As usual, nearly all the time is in llvm. So we must be generating pathological .ll 

So I'm not sure how to fix this without just not using this schedule - it scalarizes like crazy. I guess I could try to see how to avoid scalarization. It seems to be in the tail case though.

**Comment by steven-johnson** [comment]
Let me look at the revision history and see if that tells me anything.

**Comment by steven-johnson** [comment]
Adding @vksnk as he was involved in both the google-specific code and the version going into apps/nn_ops -- maybe he can shed light on whether the simpler version in apps/nn_ops came first (and the more complex version in google was a later rev), or vice versa, and why there is a divergence.

**Comment by vksnk** [comment]
Yeah, this generator was always the slowest one to compile. I think the one in google is a more recent and, if I remember correctly, has a slightly better schedule for some of the corner cases. That being said, I don't think this particular implementation is actively used right now, so probably should be okay to comment part of the schedule out to unblock the change. I can take care of it.

 Although, we still probably want to report this to QC, so they can take a look (especially that there is a public repro case).

**Comment by dsharletg** [comment]
Was this fixed by #4744?

**Comment by abadams** [comment]
> [!NOTE]
> **Automated triage:** this issue was flagged by Claude as likely safe to close, for the reason below. The reasoning was checked against the current source, and a second pass tried and failed to refute it, but no human verified this specific text. **If the reason is wrong, please reopen the issue and say so** — some fraction of these are expected to be wrong.

**Reason (CLOSE_FIXED):** The 45-minute compile came from massive scalarization of predicated loads/stores in the Hexagon path. #4744, which dsharletg pointed at in the last comment, was merged specifically to fix this exact schedule and reports bringing it down to ~2 minutes. Separately the repro itself is gone: `apps/nn_ops` was deleted in favor of `hannk`.

**Evidence:** PR #4744 "Fix pathologically slow depthwise convolution compilation" (merged 2020-03-25) touches `src/CodeGen_Hexagon.cpp`, `src/CodeGen_Internal.cpp/.h`, `src/HexagonOffload.cpp`, and its description quotes the exact schedule from this issue and states it reduces 45 min to 2 min; `apps/nn_ops` was removed in commit abefa3c43 "Remove nn_ops app in favor of hannk app. (#5893)".

Closing: #4744 (merged 2020-03-25) targeted this exact schedule — it scalarizes predicated loads/stores much more carefully in `CodeGen_Hexagon`/`CodeGen_Internal` and its description reports the compile time dropping from 45 min to ~2 min. The repro itself is also gone, since `apps/nn_ops` was removed in abefa3c43 (#5893). Please reopen with a fresh repro if pathological Hexagon compile times still show up.
